### PR TITLE
implement get_state_root() and get_fork()

### PR DESCRIPTION
**Pull request by jmcook1186**
Implements `get_state_root` and `get_fork` functions in `api_client`.

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -26,6 +26,14 @@ pub enum StateId {
     Root(Root),
 }
 
+
```
Comment (ralexstokes) [inline src/types.rs:43]:
```suggestion
```

File: src/types.rs [hunk 2]
```diff
@@ -26,6 +26,14 @@ pub enum StateId {
...
+#[derive(Serialize, Deserialize)]
+pub struct RootData {
+    pub root: Root,
+}
+
+
+
```
Comment (ralexstokes) [inline src/types.rs]:
```suggestion
```

File: src/api_client.rs [hunk 3]
```diff
@@ -111,12 +112,40 @@ impl Client {
...
+            StateId::Genesis => stub = "genesis",
+            StateId::Root(_) => stub = "not implemented",
+            StateId::Slot(_) => stub = "not implemented",
+            StateId::Justified => stub = "justified",
+        };
+
+        let path = format!("eth/v1/beacon/states/{}/root", &stub);
+
+        let root: Value<RootData> = self.get(&path).await?;
+
```
Comment (ralexstokes) [inline src/api_client.rs]:
```suggestion
```

File: src/api_client.rs [hunk 4]
```diff
@@ -111,12 +112,40 @@ impl Client {
...
+            StateId::Finalized => stub = "finalized",
+            StateId::Head => stub = "head",
+            StateId::Genesis => stub = "genesis",
+            StateId::Root(_) => stub = "not implemented",
+            StateId::Slot(_) => stub = "not implemented",
+            StateId::Justified => stub = "justified",
+        };
+
+        let path = format!("eth/v1/beacon/states/{}/root", &stub);
+
```
Comment (ralexstokes) [inline src/api_client.rs]:
```suggestion
```

File: src/api_client.rs [hunk 5]
```diff
@@ -111,12 +112,40 @@ impl Client {
...
+        match state_id {
+            StateId::Finalized => stub = "finalized",
+            StateId::Head => stub = "head",
+            StateId::Genesis => stub = "genesis",
+            StateId::Root(_) => stub = "not implemented",
+            StateId::Slot(_) => stub = "not implemented",
+            StateId::Justified => stub = "justified",
+        };
+
+        let path = format!("eth/v1/beacon/states/{}/root", &stub);
```
Comment (ralexstokes) [inline src/api_client.rs]: so a different way to do this that avoids having to copy this code each time we use it is to move the logic to the type, in this case `StateId`

we can then identify the common behavior here (aka "rendering this as a string") and implement a trait for that behavior

rust already has many built-in traits for all sorts of conversions, like ToString: https://doc.rust-lang.org/std/string/trait.ToString.html and if you read the docs then it requests you to implement `fmt::Display`

that gives an impl of `to_string` for free and the way you have set up this `format!` call it should call the right logic for you

you don't even need to pass `state_id` by reference, you can just drop it into the call to `format!`.

so I would implement `std::fmt::Display` for `StateId` and return the correct string data for each variant

then you can delete this block here and the one below

**Review by ralexstokes — CHANGES_REQUESTED** [review]
nice work!

left a few comments about making a more streamlined PR

File: src/api_client.rs (hunk 5 above)
Comment (jmcook1186) [reply src/api_client.rs]: ok, yea that seems obvious now. Fixed!

**Review by ralexstokes — CHANGES_REQUESTED** [review]
~looks like it needs `rustfmt`~

previous comment was miscopied from other pr, waiting on changes here

File: src/types.rs [hunk 6]
```diff
@@ -26,6 +27,25 @@ pub enum StateId {
...
 
+impl fmt::Display for StateId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let printable = match *self {
+            StateId::Finalized => "finalized",
+            StateId::Justified => "justified",
+            StateId::Head => "head",
+            StateId::Genesis => "genesis",
+            StateId::Slot(_) => "not yet implemented",
+            StateId::Root(_) => "not yet implemented",
```
Comment (ralexstokes) [inline src/types.rs]: you can inline the call to `write` for the `slot` and `root`

```rust
let printable = match *self {
    // others
   StateId::Slot(slot) => return write!(f, "{}", slot),
   // same for Root
};
write!(f, "{}", printable)
```

**Review by ralexstokes — CHANGES_REQUESTED** [review]
one more thing and then we merge :)

File: src/types.rs (hunk 6 above)
Comment (jmcook1186) [reply src/types.rs]: ok!

**Review by ralexstokes — APPROVED** [review]
lgtm! excellent work!!

**Pull request merged**